### ISSUE TITLE
Wrong emulation of INSERT .. VALUES .. ON DUPLICATE KEY IGNORE for multi row inserts

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8922]
The current emulation of `INSERT .. VALUES .. ON DUPLICATE KEY IGNORE` with multi value inserts is not emulated correctly when the emulation uses `MERGE`.

**jOOQ logic**:

```java
create().insertInto(TAuthor(), TAuthor_ID(), TAuthor_LAST_NAME())
        .values(1, "A")
        .values(2, "B")
        .values(3, "C")
        .values(4, "D")
        .onDuplicateKeyIgnore()
        .execute());
```

**Generated SQL** (e.g. SQL Server):

```sql
merge into [test].[dbo].[t_author]
using (
  select 1 [one]
) t
on (
  [test].[dbo].[t_author].[ID] = 4
  or [test].[dbo].[t_author].[LAST_NAME] = 'D'
)
when not matched then insert (
  [ID], 
  [LAST_NAME]
)
values (
  4, 
  'D'
);
```

Affected dialects are:

- DB2
- Derby
- HSQLDB
- Informix
- Oracle
- SQL Server
- Sybase SQL Anywhere
- Teradata

The emulation using `INSERT .. SELECT` is not affected

**Comment by richard-berg** [comment]
Hi Lukas, is there a way to force the other emulation?

If not, can you provide sample code for building the MERGE statement by hand?  (I know the SQL syntax, just not the Jooq syntax...)

**Comment by lukaseder** [comment]
See: https://www.jooq.org/doc/latest/manual/sql-building/sql-statements/merge-statement/

**Comment by richard-berg** [comment]
Thanks for the reply, but that doesn't show how to handle multi row (set-based) MERGE in Jooq.

In SQL Server I routinely do things like:

> MERGE INTO Table AS target
> USING @newValues AS source -- might need to stick a subselect in here for compatibility?
> ON source.id = target.id
> WHEN MATCHED AND source.field <> target.field
>   THEN UPDATE SET field = source.field
> WHEN NOT MATCHED BY TARGET 
>   THEN INSERT (id, field)
>   VALUES(source.id, source.field)

Apologies for my tone in https://github.com/jOOQ/jOOQ/issues/9340 but I'm afraid I'm back where I started: trying to convert SQL Server procs to Jooq so that they also work in Postgres and HSQLDB. 
 The Oracle-style "DUAL" syntax in the manual is sufficiently foreign to me that I don't know how to match it up with the constructs expected by Jooq.

I assume I need to wire up a table-valued param for the former, and a subselect for the latter two?  Where?

**Comment by lukaseder** [comment]
We've just added supporting multiple `WHEN [ NOT ] MATCHED` clauses, and the `AND` sub-clause, targeting for jOOQ 3.14: https://github.com/jOOQ/jOOQ/issues/7291. You can see how to emulate the clause for the time being in the same issue.

I'll comment on individual points:

> USING @newValues AS source -- might need to stick a subselect in here for compatibility?

Yes, a subselect works better for multiple databases - I'd say it's even better in SQL Server from a performance perspective, unless SQL Server can inline table variables instead of buffering them eagerly?

> WHEN MATCHED AND source.field <> target.field

The `AND` clause can be emulated using a `CASE` expression, see issue above. Starting from jOOQ 3.14 (early access here: https://www.jooq.org/download/versions), you can use it natively in jOOQ.

> WHEN NOT MATCHED BY TARGET

`BY TARGET` is just for clarification / consistency, it can be omitted as well in SQL Server, to my understanding. The `BY SOURCE` clause will be supported in jOOQ, soon, although it will probably be difficult to emulate: #10059

> trying to convert SQL Server procs to Jooq so that they also work in Postgres and HSQLDB.

This is a complicated area, and jOOQ will not be the silver bullet you might be expecting here. PostgreSQL doesn't support `MERGE`, it only supports `ON CONFLICT`. You could use `ON CONFLICT` on all three dialects, as jOOQ can emulate it using `MERGE`. The other way round is not possible.

The blog post I've linked from the other issue (https://blog.jooq.org/2020/04/10/the-many-flavours-of-the-arcane-sql-merge-statement/) hints at how different the various flavours of `MERGE` are across database products. And this is just about SQL syntax and functionality. The post deliberately does not touch on topics such as performance or locking semantics...

> The Oracle-style "DUAL" syntax in the manual is sufficiently foreign to me that I don't know how to match it up with the constructs expected by Jooq.

It is not really significant. Most database dialects require a `FROM` clause, and as such, offer a dummy table, such as `DUAL`. SQL Server does not have this requirement, it is possible to generate a single row using `SELECT 1`.

> I assume I need to wire up a table-valued param for the former, and a subselect for the latter two? Where?

I think, given your constraints of having to support the three dialects you've mentioned, you will have to find a more generic solution than the SQL Server specific `MERGE` statement that you would like to write. This isn't strictly a jOOQ question, but a generic dialect interoperability question. `INSERT .. ON CONFLICT` might help here, though the query will be complex.

Despite of using jOOQ, it is not entirely forbidden to write 2% - 5% vendor specific queries, even in stored procedures (and call those using jOOQ). Just because you're using jOOQ doesn't mean you have to use jOOQ for everything.

**Comment by lukaseder** [comment]
This has been fixed in jOOQ 3.14 with https://github.com/jOOQ/jOOQ/issues/10461. Closing as duplicate
